### PR TITLE
Feat/similarity

### PR DESCRIPTION
**Pull request by melisandequentzel**
Add snippet_similarity.py + add update_similar_snippets method to client_sqlite.py + update requirements.txt

### CODE REVIEW

**Comment by marcorosa** [comment]
Hi @melisandequentzel thanks for the PR
Please change the base branch with `develop` instead of `main`

Then please follow all the indications of @pep8speaks that you see above to fix the style of the code. Then I will go into all the changes 😄

**Comment by melisandequentzel** [comment]
Hi @marcorosa thanks for the remarks, will make the changes to the style shortly :)

**Comment by melisandequentzel** [comment]
@marcorosa done, the last one: Line 2:1: F401 'tensorflow_text' imported but unused 
needs to remain because without that import some errors are raised

File: requirements.txt [hunk 1]
```diff
@@ -14,3 +14,6 @@ requests
 scikit-learn
 srsly
 tqdm
+tensorflow==2.4.1
```
Comment (marcorosa) [inline requirements.txt:17]: please add the dependencies in alphabetical order

File: credentialdigger/snippet_similarity.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import tensorflow as tf
+import tensorflow_text
+import logging
+import tensorflow_hub as hub
+import tensorflow.keras.preprocessing.text
+import numpy as np
```
Comment (marcorosa) [inline credentialdigger/snippet_similarity.py:6]: The order of the imports is wrong (you can use [`isort`](https://pycqa.github.io/isort/) to do it automatically)

File: credentialdigger/snippet_similarity.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import tensorflow_text
+import logging
+import tensorflow_hub as hub
+import tensorflow.keras.preprocessing.text
+import numpy as np
+
+logging.getLogger('tensorflow').setLevel(logging.ERROR)
+
+
+def build_embedding_model():
```
Comment (marcorosa) [inline credentialdigger/snippet_similarity.py:11]: missing documentation

File: credentialdigger/snippet_similarity.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def compute_snippet_embedding(snippet, model):
+    small_bert_result = tf.squeeze(model(tf.constant(snippet)))
+    small_bert_embeddings = small_bert_result.numpy()
+    snippet_embedding = np.mean(small_bert_embeddings, axis=0)
+    return snippet_embedding
+
+
+def compute_similarity(emb_1, emb_2):
```
Comment (marcorosa) [inline credentialdigger/snippet_similarity.py:35]: same as above

File: credentialdigger/client.py [hunk 5]
```diff
@@ -1006,3 +1010,25 @@ def _get_scan_rules(self, category=None, exclude=None):
             raise ValueError('No rules found')
 
         return rules
+
+    def update_similar_snippets(self,
+                                target_snippet,
+                                state,
+                                repo_url,
+                                file_name=None,
```
Comment (marcorosa) [inline credentialdigger/client.py:1018]: why do you need the filename? As far as I understood, given a `target_snippet` you need to get all the "new" discoveries for a certain repo, and then compute each similarity and eventually flag other FPs. So what's the need of having the file name in this process?

File: ui/backend/client_ui.py [hunk 6]
```diff
@@ -97,3 +101,26 @@ def check_repo(self, repo_url, git_token=None, local_repo=False):
...
+
+    def update_similar_snippets(self,
+                                target_snippet,
+                                state,
+                                repo_url,
+                                file_name=None,
+                                threshold=0.96):
+        discoveries = self.get_discoveries(repo_url,
+                                           file_name,
+                                           state_filter='new')[1]
```
Comment (marcorosa) [inline ui/backend/client_ui.py:113]: why do you only keep the [1] instead of the whole list?
are you sure you don't raise an IndexError if you don't have other discoveries?

File: ui/backend/client_ui_sqlite.py [hunk 7]
```diff
@@ -87,6 +91,9 @@ def get_discoveries(self, repo_url, file_name=None, state_filter=None,
         # Build outer query to get all occurrences of the paginated snippets
         query = 'SELECT * FROM discoveries WHERE repo_url=?'
         params = [repo_url]
+        params = [repo_url]
+        params = [repo_url]
+        params = [repo_url]
```
Comment (marcorosa) [inline ui/backend/client_ui_sqlite.py:96]: ?

File: ui/backend/client_ui_sqlite.py [hunk 8]
```diff
@@ -165,3 +172,26 @@ def get_files_summary(self, repo_url):
                 " FROM discoveries WHERE repo_url=?"
                 " GROUP BY file_name"
             ))
+
+    def update_similar_snippets(self,
```
Comment (marcorosa) [inline ui/backend/client_ui_sqlite.py:176]: why do you need to redefine this (or to copy it from the other class)?

**Comment by marcorosa** [comment]
Hi, apart from all the changes/doubts requested above, I have 2 further concerns?
- if/when the PR is ready to be reviewed, remember to turn it to "ready for review" and don't leave it draft
- There are few comments, and the code is hard to follow. Please write comments in the code, and remember to write the doctrings for ALL your new methods ;)

File: requirements.txt [hunk 9]
```diff
@@ -14,3 +14,6 @@ requests
...
+tensorflow-text==2.4.1
+tf-models-official
```
[the remark this answers is not in the document]
Comment (melisandequentzel) [reply requirements.txt:19]: Version 2.4.1 isn't available for tf-models-official, the right version is picked up automatically. The rule to have matching versions only applies to tf and tf-text

File: credentialdigger/client.py (hunk 5 above)
Comment (melisandequentzel) [reply credentialdigger/client.py:1018]: The purpose of having a filename here is to allow for exclusive updating of similar snippets found in the same file as the target snippet, because there could be a situation where a file contains similar snippets as another file, but one with real passwords and the other with dummy ones. This parameter is provided 'just in case' someone needs this feature, but is not meant to be used as a general rule. I can remove it if you think it should not be there.

File: ui/backend/client_ui.py (hunk 6 above)
Comment (melisandequentzel) [reply ui/backend/client_ui.py:113]: I added [1] because the get_discoveries method from the UI classes returns the number of discoveries as first element, which is useless in this case. I'm only retriving the actual discoveries, not their count. This is what adding [1] allows for.

File: ui/backend/client_ui_sqlite.py (hunk 8 above)
Comment (melisandequentzel) [reply ui/backend/client_ui_sqlite.py:176]: Precisely because of the get_discoveries method that is specific to the UI classes. No other way to go than to rewrite the adapted version of update_similar_snippets in the UI classes

Comment (melisandequentzel) [reply ui/backend/client_ui_sqlite.py:176]: Or maybe checking for length of method output directly in client.py, but doesn't seem any more efficient

**Comment by melisandequentzel** [comment]
Hi, okay, thank you for the reviews. I will add comments + docstrings, and will make sure to do so and to have the right order for imports in the future :)

File: credentialdigger/client.py (hunk 5 above)
Comment (marcorosa) [reply credentialdigger/client.py:1016]: Let's keep it with this PR then, but before merging into the main branch I would like to test it. Moreover, I'm not fully convinced this parameter is really needed: if the same snippet appears in 2 different files, then either they are treated at the same way (i.e., both TP or both FP), or it is the path model that should flag one of the two as a FP.

File: ui/backend/client_ui_sqlite.py [hunk 10]
```diff
@@ -165,3 +169,26 @@ def get_files_summary(self, repo_url):
...
+                                target_snippet,
+                                state,
+                                repo_url,
+                                file_name=None,
+                                threshold=0.95):
```
Comment (marcorosa) [inline ui/backend/client_ui_sqlite.py:178]: it was 0.96 in the parent class. Why have you changed it here?

File: credentialdigger/client.py [hunk 11]
```diff
@@ -1006,3 +1008,51 @@ def _get_scan_rules(self, category=None, exclude=None):
...
+            if d['state'] == 'new':
+                """ Compute snippet embedding """
+                snippet_embedding = compute_snippet_embedding(d['snippet'],
+                                                              model)
+                """ Compute similarity of target snippet and snippet """
+                similarity = compute_similarity(target_snippet_embedding,
+                                                snippet_embedding)
+                if similarity > threshold:
+                    n_updated_snippets += 1
+                    self.update_discovery(d['id'], state)
```
Comment (marcorosa) [inline credentialdigger/client.py:1057]: does it work without passing the query like in the update_discovery method?

**Review by marcorosa — CHANGES_REQUESTED** [review]
You are missing the client_ui_postgres class

File: credentialdigger/client.py (hunk 5 above)
Comment (melisandequentzel) [reply credentialdigger/client.py:1016]: Yes in theory, but then why need a snippet model if the path model is always right? Some snippets with fake credentials can appear in paths that aren't dedicated to testing and/or detected as dummy by the path model, since we're applying a model to distinguish real from fake passwords after the path model. The snippet model may fail to recognize a fake password as being fake, and that's the whole point of having a similarity model: to provide a simple, 'user-powered' way of compensating for failings/shortcomings of one of the two models. This may become superfluous when the snippet model reaches a very high accuracy, but seems quite relevant as of now. Adding the filename parameter offers a more granular control over the update, which can't be a bad thing. In the UI, it could just take the form of a button to tick 'restrict updates to current file'. I noticed, there are often similar snippets in the same file, so it would still be efficient using it that way.

Comment (melisandequentzel) [reply credentialdigger/client.py:1016]: (There clearly is a (very) small danger bound to the similarity feature, and steps to mitigate this danger should be taken, I think. Otherwise, we'll need to clearly inform the user on the potential risks of overly using this functionality on snippets that do contain a password).

File: credentialdigger/client.py [hunk 12]
```diff
@@ -1006,3 +1008,51 @@ def _get_scan_rules(self, category=None, exclude=None):
...
+        ----------
+        target_snippet: str
+        state: str
+            state to update similar snippets to
+        repo_url: str
+        file_name: str
+            restrict to a given file the search for similar snippets
+        threshold: double
+            update snippets with similarity score above threshold.
+            Values lesser than 0.94 do not generally imply any relevant
```
[the remark this answers is not in the document]
Comment (melisandequentzel) [reply credentialdigger/client.py:1031]: 0.94, it's intended as a guideline on which values to use for this parameter. 0.96 is usually perfect, but in some rare cases, putting a lower value will ensure all similar snippets will be detected; however that value should always be greater than 0.94.

File: ui/backend/client_ui.py [hunk 13]
```diff
@@ -97,3 +101,49 @@ def check_repo(self, repo_url, git_token=None, local_repo=False):
...
+        ----------
+        target_snippet: str
+        state: str
+            state to update similar snippets to
+        repo_url: str
+        file_name: str
+            restrict to a given file the search for similar snippets
+        threshold: double
+            update snippets with similarity score above threshold.
+            Values lesser than 0.94 do not generally imply any relevant
```
[the remark this answers is not in the document]
Comment (melisandequentzel) [reply ui/backend/client_ui.py:124]: 0.94

File: ui/backend/client_ui_sqlite.py (hunk 10 above)
Comment (melisandequentzel) [reply ui/backend/client_ui_sqlite.py:178]: Indeed, I forgot to update the value here, I thought I had done so. It should be 0.96

File: credentialdigger/client.py (hunk 11 above)
Comment (melisandequentzel) [reply credentialdigger/client.py:1057]: Yes, because the update_discovery method that gets called is that of the child class, which doesn't require a query parameter. It definitely works.

File: ui/backend/client_ui_sqlite.py (hunk 10 above)
Comment (melisandequentzel) [reply ui/backend/client_ui_sqlite.py:178]: Oh but the code should have been removed from that class altogether, that's why I didn't update it. I'm removing it now

**Comment by melisandequentzel** [comment]
@marcorosa I'm not missing the code for the client_ui_postgres file because it's in the parent class (client_ui file). I had failed to remove the code from the client_ui _sqlite file. I just did so, and re-tested it.

**Comment by melisandequentzel** [comment]
@marcorosa I fixed the float types and removed code from client_ui_sqlite, so the code should be ready by now. I replied to your comment regarding the filename parameter in the update_similar_snippets method. Should I make the change anyway?

**Review by marcorosa — APPROVED** [review]
